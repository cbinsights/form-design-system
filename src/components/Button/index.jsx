import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from 'util/baseElement';
import { CaretDownIcon } from 'lib/icons/react';

export const SIZES = ['s', 'm'];
export const THEMES = ['blue', 'outlined', 'ghost'];

const Button = ({
  theme = 'blue',
  size = 'm',
  IconLeft,
  IconRight,
  Link,
  disabled,
  label,
  isActive,
  isLoading,
  isDestructive,
  isFullWidth,
  hasCaret,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });

  const IconComponent = ({ direction }) => {
    const Icon = direction === 'left' ? IconLeft : IconRight;
    return (
      <>
        {Icon && (
          <div
            className={cx('alignChild--center--center', `fdsButton-icon--${direction}`, {
              'fdsButton--hidden': isLoading,
            })}
          >
            <Icon customSize={size === 's' ? 16 : 18} />
          </div>
        )}
      </>
    );
  };

  IconComponent.propTypes = {
    direction: PropTypes.oneOf(['left', 'right']),
  };

  const Icons = () => (
    <>
      <IconComponent direction="left" />
      <IconComponent direction="right" />
    </>
  );

  return (
    <Element
      {...rest}
      className={cx(
        'fdsButton',
        'fontStyle--caps',
        'display--inlineFlex',
        'rounded--all',
        'alignChild--center--center',
        'border--focus--noTransition',
        'transition--default',
        `fdsButton--${size}`,
        `fdsButton--${theme}`,
        {
          'fdsButton--isDestructive': isDestructive,
          'fdsButton--loading': isLoading,
          'fdsButton--disabled': disabled,
          'fdsButton--isFullWidth': isFullWidth,
          'fdsButton--isActive': isActive && !disabled,
        }
      )}
      disabled={disabled && Element === 'button'}
    >
      <span className={cx('fdsButton-label', { 'fdsButton--hidden': isLoading })}>
        {label}
      </span>
      <Icons />
      {hasCaret && (
        <div className="margin--left--half alignChild--center--center">
          <CaretDownIcon customSize={12} />
        </div>
      )}
    </Element>
  );
};

Button.propTypes = {
  /** Text inside the button */
  label: PropTypes.string.isRequired,
  /** Used to control the display and theme of the button */
  theme: PropTypes.oneOf(THEMES),
  /** Used to control the size of the button */
  /**
   * Controls whether button should be in a destructive UI state.
   * This only influences UI, there is no functional differences enabled
   * with this boolean
   */
  isDestructive: PropTypes.bool,
  size: PropTypes.oneOf(SIZES),
  /**  Pass in "only" a FDS Icon reference to display it (e.g. IconRight={ApproveIcon}) */
  IconRight: PropTypes.func,
  /**  Pass in "only" a FDS Icon reference to display it (e.g. IconLeft={ApproveIcon}) */
  IconLeft: PropTypes.func,
  /** Controls showing CaretDown icon (right aligned) */
  hasCaret: PropTypes.bool,
  /**
   * Controls whether the button is disabled or not. Will control the disabled
   * presentation of either an anchor or button rendered under the hood,
   * but will only add a disabled attribute for buttons
   */
  disabled: PropTypes.bool,
  /** Controls the button going full width */
  isFullWidth: PropTypes.bool,
  /** Controls showing of an active UI state for the button */
  isActive: PropTypes.bool,
  /**
   * Controls whether loading spinner displays. Button text and icons are set to visibility hidden
   * to preserve the space set, whilst hiding them
   */
  isLoading: PropTypes.bool,
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
};

export default Button;
