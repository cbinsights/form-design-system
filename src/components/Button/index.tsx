import React, { forwardRef, HTMLAttributes } from 'react';
import cc from 'classcat';
import baseElement from 'util/baseElement';
import { CaretDownIcon } from 'icons/react';

export const SIZES = ['s', 'm'] as const;

export const THEMES = ['blue', 'orange', 'outlined', 'outlined--blue', 'ghost'] as const;
export type Theme = typeof THEMES[number];

export interface ButtonProps
  extends HTMLAttributes<HTMLElement | HTMLButtonElement | HTMLAnchorElement> {
  /** Text inside the button */
  label: string;

  /** Used to control the display and theme of the button */
  theme?: Theme;

  /**
   * Controls whether button should be in a destructive UI state.
   * This only influences UI, there is no functional differences enabled
   * with this boolean
   */
  isDestructive?: boolean;

  /** Used to control the size of the button */
  size?: typeof SIZES[number];

  /**  Pass in "only" a FDS Icon reference to display it (e.g. IconRight={ApproveIcon}) */
  IconRight?: any; // Not sure how to type this yet

  /**  Pass in "only" a FDS Icon reference to display it (e.g. IconLeft={ApproveIcon}) */
  IconLeft?: any; // Not sure how to type this yet

  /** Controls showing CaretDown icon (right aligned) */
  hasCaret?: boolean;

  /**
   * Controls whether the button is disabled or not. Will control the disabled
   * presentation of either an anchor or button rendered under the hood,
   * but will only add a disabled attribute for buttons
   */
  disabled?: boolean;

  /** Controls the button going full width */
  isFullWidth?: boolean;

  /** Controls showing of an active UI state for the button */
  isActive?: boolean;
  /**
   * Controls whether loading spinner displays. Button text and icons are set to visibility hidden
   * to preserve the space set, whilst hiding them
   */
  isLoading?: boolean;

  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: any; // Not sure how to type this yet

  /** Extend click radius of button to nearest relative parent */
  isBreakoutLink?: boolean;

  href?: string;
}

export const Button = forwardRef(
  (
    {
      theme = 'blue',
      size = 'm',
      IconLeft,
      IconRight,
      Link,
      disabled,
      label,
      isActive,
      isLoading,
      isDestructive,
      isFullWidth,
      hasCaret,
      isBreakoutLink,
      href,
      ...rest
    }: ButtonProps,
    ref
  ) => {
    const Element = baseElement({ href, onClick: true, as: Link });
    interface IconProps {
      direction?: 'left' | 'right';
    }

    const IconComponent = ({ direction }: IconProps) => {
      const Icon = direction === 'left' ? IconLeft : IconRight;
      return (
        <>
          {Icon && (
            <div
              className={cc([
                {
                  'fdsButton--hidden': isLoading,
                },
                'alignChild--center--center',
                `fdsButton-icon--${direction}`,
              ])}
            >
              <Icon customSize={size === 's' ? 16 : 18} />
            </div>
          )}
        </>
      );
    };

    const Icons = () => (
      <>
        <IconComponent direction="left" />
        <IconComponent direction="right" />
      </>
    );

    return (
      <Element
        {...rest}
        ref={ref}
        className={cc([
          {
            'fdsButton--isDestructive': isDestructive,
            'fdsButton--loading': isLoading,
            'fdsButton--disabled': disabled,
            'fdsButton--isFullWidth': isFullWidth,
            'fdsButton--isActive': isActive && !disabled,
            breakoutLink: isBreakoutLink,
          },
          'fdsButton',
          'display--inlineFlex',
          'rounded--all',
          'alignChild--center--center',
          'border--focus--noTransition',
          'transition--default',
          `fdsButton--${size}`,
          `fdsButton--${theme}`,
        ])}
        disabled={disabled && Element === 'button'}
        href={href}
      >
        <div className="display--inlineFlex fdsButton--loading-spinnerWrapper">
          <div className={cc({ 'fdsButton--loading-spinner': isLoading })} />
          <span className={cc([{ 'fdsButton--hidden': isLoading }, 'fdsButton-label'])}>
            {label}
          </span>
          <Icons />
          {hasCaret && (
            <div className="margin--left--s alignChild--center--center">
              <CaretDownIcon customSize={12} />
            </div>
          )}
        </div>
      </Element>
    );
  }
);

Button.displayName = 'Button';

export default Button;
