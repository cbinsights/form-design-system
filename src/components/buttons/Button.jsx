import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../../util/baseElement';
import CaretDownIcon from '../../../lib/icons/react/CaretDownIcon';

export const SIZES = ['s', 'm'];
export const THEMES = ['blue', 'outlined', 'ghost'];
export const ICON_PLACEMENTS = ['left', 'right'];

const Button = ({
  theme,
  iconPlacement,
  Icon,
  Link,
  size,
  disabled,
  children,
  isLoading,
  isDestructive,
  isFullWidth,
  hasCaret,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });

  return (
    <Element
      {...rest}
      className={cx(
        'fdsButton',
        'fontStyle--caps',
        'display--inlineFlex',
        'rounded--all',
        'alignChild--center--center',
        'border--focus--noTransition',
        'transition-default',
        {
          'fdsButton--s': size === 's',
          'fdsButton--m': size === 'm',
          'fdsButton--blue': theme === 'blue',
          'fdsButton--outlined': theme === 'outlined',
          'fdsButton--ghost': theme === 'ghost',
          'fdsButton--isDestructive': isDestructive,
          'fdsButton--loading': isLoading,
          'fdsButton--disabled': disabled,
          'fdsButton--isFullWidth': isFullWidth,
        }
      )}
      disabled={disabled && Element === 'button'}
    >
      <span className={cx('fdsButton-label', { 'fdsButton--hidden': isLoading })}>
        {children}
      </span>
      {Icon && (
        <div
          className={cx('alignChild--center--center', {
            'fdsButton-icon--left': iconPlacement === 'left',
            'fdsButton-icon--right': iconPlacement === 'right',
            'fdsButton--hidden': isLoading,
          })}
        >
          <Icon customSize={size === 's' ? 16 : 18} />
        </div>
      )}
      {hasCaret && (
        <div className="margin--left--half alignChild--center--center">
          <CaretDownIcon customSize={12} />
        </div>
      )}
    </Element>
  );
};

Button.defaultProps = {
  theme: 'blue',
  iconPlacement: 'right',
  size: 'm',
};

Button.propTypes = {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
  /**
   * Controls whether loading spinner displays. Button text and icons are set to visibility hidden
   * to preserve the space set, whilst hiding them
   */
  isLoading: PropTypes.bool,
  /**
   * Controls whether button should be in a destructive UI state.
   * This only influences UI, there is no functional differences enabled
   * with this boolean
   */
  isDestructive: PropTypes.bool,
  /**
   * Controls whether the button is disabled or not. Will control the disabled
   * presentation of either an anchor or button rendered under the hood,
   * but will only add a disabled attribute for buttons
   */
  disabled: PropTypes.bool,
  /** Controls which side the `Icon` renders on, assuming you pass it */
  iconPlacement: PropTypes.oneOf(ICON_PLACEMENTS),
  /**  Pass in "only" a FDS Icon reference to display it (e.g. Icon={ApproveIcon}) */
  Icon: PropTypes.func,
  /** Controls the button going full width */
  isFullWidth: PropTypes.bool,
  /** Controls showing CaretDown icon (right aligned) */
  hasCaret: PropTypes.bool,
  /** Used to control the display and theme of the button */
  theme: PropTypes.oneOf(THEMES),
  /** Used to control the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Contents inside the button */
  children: PropTypes.node,
};

export default Button;
