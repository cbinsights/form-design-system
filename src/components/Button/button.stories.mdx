import { Meta, Preview, Story, Props } from "@storybook/addon-docs/blocks";
import {
  withKnobs,
  text,
  boolean,
  optionsKnob as options
} from "@storybook/addon-knobs";
import { arrayToOptions, StoryWrapper, StoryItem, ImportPath, State } from "components/util/storybook";
import Button, { THEMES } from "./";
import { IconButton } from "components";
import StarFilledIcon from "lib/icons/react/StarFilledIcon";
import ButtonGroup from "components/ButtonGroup";

export const SIZES = ['s', 'm'];

<Meta title="Components/Button" decorators={[withKnobs]} component={Button} />

# Button

<ImportPath component={Button} />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <div className={boolean("Inverted", false) ? 'inverted' : ''}>
    <Button
      label={text("text", "Button")}
      size={options("size", arrayToOptions(SIZES), undefined, {
        display: "inline-radio"
      })}
      theme={options(
        "theme",
        arrayToOptions(THEMES),
        undefined,
        {
          display: "inline-radio"
        }
      )}
      isDestructive={boolean("isDestructive", undefined)}
      IconLeft={boolean("IconLeft") ? StarFilledIcon : undefined}
      IconRight={boolean("IconRight") ? StarFilledIcon : undefined}
      isActive={boolean("isActive", undefined)}
      disabled={boolean("disabled", undefined)}
      isLoading={boolean("isLoading", undefined)}
      hasCaret={boolean("hasCaret", undefined)}
      isFullWidth={boolean("isFullWidth", undefined)}
      isBreakoutLink={boolean("isBreakoutLink", undefined)}
    />
  </div>
</Story>

<Preview isExpanded>
  <Button label="Button" />
</Preview>

<Props of={Button} />

### Themes

<Preview>
  <Story name="Themes">
    <StoryWrapper>
    <IconButton Icon={StarFilledIcon} label="Icon" />
    {THEMES.map((theme, idx) => (
      <StoryItem key={idx}>
        <Button
          theme={theme}
          label={theme}
        />
      </StoryItem>
    ))}
    </StoryWrapper>
  </Story>
</Preview>


### isDestructive

Modifies the theme. Should only be used when the interaction involves data loss, and should also prompt the user to confirm the action.

<Preview>
  <Story name="Destructive">
    <StoryWrapper>
    {THEMES.map(theme => (
      <StoryItem>
        <Button
          theme={theme}
          isDestructive
          label={theme}
        />
      </StoryItem>
    ))}
    </StoryWrapper>
  </Story>
</Preview>

 ### Inverted

We currently support inverting `ghost` and `outlined` theme
buttons. Simply surround the button(s) with an `.inverted` class.

<Preview>
  <Story name="Inverted">
    <div class="inverted bgColor--blue padding--all rounded--all">
      <StoryWrapper>
        <StoryItem>
          <Button theme="ghost" label="Button" />
        </StoryItem>
        <StoryItem>
          <Button theme="outlined" label="Button" />
        </StoryItem>
      </StoryWrapper>
    </div>
  </Story>
</Preview> 


### isActive

`isActive` should be used on buttons to signify the button having
some affect (applying a filter, causing a popover to be open).
**Click to toggle on and off**.

<Preview>
  <Story name="isActive">
    <>
      <StoryWrapper>
        <State
          initialValue={{}}
          render={(value, setValue) => ( 
            <>
              {THEMES.map((theme, idx) => (
                <StoryItem>
                  <Button
                    theme={theme}
                    isActive={!value[idx]}
                    label={theme}
                    onClick={() => setValue({ ...value, [idx]: !Boolean(value[idx])})}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
        <State
          initialValue={{}}
          render={(value, setValue) => ( 
            <>
              {THEMES.map((theme, idx) => (
                <StoryItem>
                  <Button
                    theme={theme}
                    isActive={!value[idx]}
                    isDestructive
                    label={theme}
                    onClick={() => setValue({ ...value, [idx]: !Boolean(value[idx])})}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
      </StoryWrapper>
      <div class="inverted bgColor--blue padding--all rounded--all margin--top">
        <StoryWrapper>
          <State
            initialValue={{}}
            render={(value, setValue) => ( 
              <>
                {['ghost', 'outlined'].map((theme, idx) => (
                  <StoryItem>
                    <Button
                      theme={theme}
                      isActive={!value[idx]}
                      label={theme}
                      onClick={() => setValue({ ...value, [idx]: !Boolean(value[idx])})}
                    />
                  </StoryItem>
                ))}
              </>
            )}
          />
        </StoryWrapper>
      </div>
    </>
  </Story>
</Preview>

### Sizes

<Preview>
  <Story name="Button Sizes">
    {SIZES.map(size => (
      <div className="display--inlineBlock margin--right--half margin--bottom--half">
        <Button
          size={size}
          label="Button"
        />
      </div>
    ))}
  </Story>
</Preview>

### Full width button

<Preview>
  <Story name="Full Width">
    <Button
      isFullWidth
      label="Button"
    />
  </Story>
</Preview>

### Icons

ONLY pass in FDS Icon references to `IconLeft` or `IconRight`. Very important. (refer to code sample)

<Preview>
  <Story name="Icons">
    <StoryWrapper>
      <StoryItem>
        <Button IconLeft={StarFilledIcon} label="Button" />
      </StoryItem>
      <StoryItem>
        <Button IconRight={StarFilledIcon} label="Button" />
      </StoryItem>
      <StoryItem>
        <IconButton Icon={StarFilledIcon} label="Button" />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

### Caret

<Preview>
  <Story name="Caret">
    <StoryWrapper>
      <StoryItem>
        <Button theme="blue" hasCaret label="Caret" />
      </StoryItem>
      <StoryItem>
        <Button theme="blue" hasCaret IconRight={StarFilledIcon} label="Caret" />
      </StoryItem>
      <StoryItem>
        <Button theme="blue" hasCaret IconRight={StarFilledIcon} label="Caret" />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

### Loading

Loading states should be shown whenever an action has the ability 
to take more than a couple of seconds of loading time

<Preview>
  <Story name="Loading">
    <>
      <StoryWrapper>
      {THEMES.map(theme => (
        <StoryItem>
          <Button
            theme={theme}
            isLoading
            label="Button"
          />
        </StoryItem>
      ))}
      {THEMES.map(theme => (
        <StoryItem>
          <Button
            theme={theme}
            isDestructive
            isLoading
            label="Button"
          />
        </StoryItem>
      ))}
      </StoryWrapper>
      <div class="inverted bgColor--blue padding--all rounded--all margin--top">
        <StoryWrapper>
          <StoryItem>
            <Button theme="ghost" isLoading label="Button" />
          </StoryItem>
          <StoryItem>
            <Button theme="outlined" isLoading label="Button" />
          </StoryItem>
        </StoryWrapper>
      </div>
    </>
  </Story>
</Preview>

### Breakout

<Preview>
  <Story name="Breakout">
    <StoryWrapper>
      <StoryItem>
        <div className="alignChild--center--center" style={{ width: '100px', height: "100px", border: "1px dashed #f09", position: "relative" }}>
          <Button label="Button" onClick={() => console.log('hey')} isBreakoutLink />
        </div>
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

### Misc

<Preview>
  <Story name="Misc">
    <StoryWrapper>
      <StoryItem>
        <div style={{ width: '100px' }}>
          <Button label="Text can wrap" />
        </div>
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>
