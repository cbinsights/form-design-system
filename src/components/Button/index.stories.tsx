import React from 'react';
import State from 'util/storybook-docs/State';
import { StoryObj } from '@storybook/react';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import { StarFilledIcon, WorldIcon } from 'icons/react';
import Button, { ButtonProps, THEMES, Theme, SIZES } from '.';

const INVERTED_THEMES: Theme[] = ['ghost', 'outlined'];

export const Primary: StoryObj<ButtonProps> = {
  render: ({ label = 'Button', ...args }: ButtonProps) => (
    <Button {...args} label={label} />
  ),
  args: {
    label: 'Button',
  },
};

export const Themes: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <StoryWrapper>
      {THEMES.map((theme, idx) => (
        <StoryItem key={idx}>
          <Button {...args} theme={theme} label={theme} />
        </StoryItem>
      ))}
    </StoryWrapper>
  ),
};

export const IsDestructive: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <StoryWrapper>
      {THEMES.map((theme, idx) => (
        <StoryItem key={idx}>
          <Button {...args} isDestructive theme={theme} label={theme} />
        </StoryItem>
      ))}
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Modifies the theme. Should only be used when the interaction involves data loss, and should also prompt the user to confirm the action.',
      },
    },
  },
};

export const Inverted: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <div className="inverted bgColor--blue padding--all rounded--all">
      <StoryWrapper>
        <StoryItem>
          <Button {...args} theme="ghost" label="Button" />
        </StoryItem>
        <StoryItem>
          <Button {...args} theme="outlined" label="Button" />
        </StoryItem>
      </StoryWrapper>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'We currently support inverting `ghost` and `outlined` theme buttons. Simply surround the button(s) with an `.inverted` class.',
      },
    },
  },
};

type Shape = {
  [key: number]: boolean;
};
const initialValue: Shape = {};

export const IsActive: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <>
      <StoryWrapper>
        <State
          initialValue={initialValue}
          render={(value, setValue) => (
            <>
              {THEMES.map((theme, idx) => (
                <StoryItem key={idx}>
                  <Button
                    {...args}
                    theme={theme}
                    isActive={!value[idx]}
                    label={theme}
                    onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
        <State<Shape>
          initialValue={{}}
          render={(value, setValue) => (
            <>
              {THEMES.map((theme, idx) => (
                <StoryItem key={idx}>
                  <Button
                    theme={theme}
                    isActive={!value[idx]}
                    isDestructive
                    label={theme}
                    onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
      </StoryWrapper>
      <div className="inverted bgColor--blue padding--all rounded--all margin--top">
        <StoryWrapper>
          <State
            initialValue={initialValue}
            render={(value, setValue) => (
              <>
                {INVERTED_THEMES.map((theme, idx) => (
                  <StoryItem key={idx}>
                    <Button
                      theme={theme}
                      isActive={!value[idx]}
                      label={theme}
                      onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                    />
                  </StoryItem>
                ))}
              </>
            )}
          />
        </StoryWrapper>
      </div>
    </>
  ),
  parameters: {
    docs: {
      description: {
        story:
          '`isActive` should be used on buttons to signify the button having some affect (applying a filter, causing a popover to be open). **Click to toggle on and off**.',
      },
    },
  },
};

export const Sizes: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <>
      {SIZES.map((size, idx) => (
        <div
          key={idx}
          className="display--inlineBlock margin--right--s margin--bottom--s"
        >
          <Button {...args} size={size} label="Button" />
        </div>
      ))}
    </>
  ),
};

export const FullWidth: StoryObj<ButtonProps> = {
  render: ({ label = 'Button', ...args }: ButtonProps) => (
    <Button {...args} label={label} />
  ),
};

export const Icons: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <Button {...args} IconLeft={WorldIcon} label="Button" />
      </StoryItem>
      <StoryItem>
        <Button {...args} IconRight={StarFilledIcon} label="Button" />
      </StoryItem>
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'ONLY pass in FDS Icon references to `IconLeft` or `IconRight`. Very important. (refer to code sample)',
      },
    },
  },
};

export const Caret: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <Button {...args} theme="blue" hasCaret label="Caret" />
      </StoryItem>
      <StoryItem>
        <Button
          {...args}
          theme="blue"
          hasCaret
          IconRight={StarFilledIcon}
          label="Caret"
        />
      </StoryItem>
      <StoryItem>
        <Button
          {...args}
          theme="blue"
          hasCaret
          IconRight={StarFilledIcon}
          label="Caret"
        />
      </StoryItem>
    </StoryWrapper>
  ),
};

export const Loading: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <>
      <StoryWrapper>
        {THEMES.map((theme, idx) => (
          <StoryItem key={idx}>
            <Button {...args} theme={theme} isLoading label="Button" />
          </StoryItem>
        ))}
        {THEMES.map((theme, idx) => (
          <StoryItem key={idx}>
            <Button {...args} theme={theme} isDestructive isLoading label="Button" />
          </StoryItem>
        ))}
      </StoryWrapper>
      <div className="inverted bgColor--blue padding--all rounded--all margin--top">
        <StoryWrapper>
          <StoryItem>
            <Button {...args} theme="ghost" isLoading label="Button" />
          </StoryItem>
          <StoryItem>
            <Button {...args} theme="outlined" isLoading label="Button" />
          </StoryItem>
        </StoryWrapper>
      </div>
    </>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Loading states should be shown whenever an action has the ability to take more than a couple of seconds of loading time',
      },
    },
  },
};

export const Breakout: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <div
          className="alignChild--center--center padding--all--xl"
          style={{
            border: '1px dashed #f09',
            position: 'relative',
          }}
        >
          <Button {...args} label="Button" isBreakoutLink />
        </div>
      </StoryItem>
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Breakout buttons extend their clickable area to the next nearest parent.',
      },
    },
  },
};

export const Misc: StoryObj<ButtonProps> = {
  render: (args: ButtonProps): JSX.Element => (
    <div
      style={{
        width: '100px',
      }}
    >
      <Button {...args} label="Text can wrap" />
    </div>
  ),
};

export default {
  component: Button,
  title: 'components/Button',
  parameters: {
    docs: {
      fds: {
        attributes: ['rest'],
      },
    },
  },
};
