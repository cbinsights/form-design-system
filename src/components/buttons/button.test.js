import React from 'react';
import { shallow } from 'enzyme';

import Button from './Button';

const renderComponent = (props) => shallow(<Button {...props} />);

const buttons = {
  Link: () => {}, isLoading: true, isDestructive: true, disabled: true, iconPlacement: 'left',
  Icon: () => {}, isFullWidth: true, theme: 'outlined', hasCaret: true
};

describe('ButtonGroup component', () => {

  it('matches snapshot (default props)', () => {
    const component = renderComponent({ children: 'Button' });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent(buttons);
    expect(component).toMatchSnapshot();
  });
});
