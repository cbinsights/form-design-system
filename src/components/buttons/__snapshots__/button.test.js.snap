// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup component matches snapshot (default props) 1`] = `
<button
  className="fdsButton fontStyle--caps display--inlineFlex rounded--all alignChild--center--center border--focus--noTransition transition-default fdsButton--m fdsButton--blue"
>
  <span
    className="fdsButton-label"
  >
    Button
  </span>
</button>
`;

exports[`ButtonGroup component matches snapshot (set all props) 1`] = `
<Link
  className="fdsButton fontStyle--caps display--inlineFlex rounded--all alignChild--center--center border--focus--noTransition transition-default fdsButton--m fdsButton--outlined fdsButton--isDestructive fdsButton--loading fdsButton--disabled fdsButton--isFullWidth"
  disabled={false}
>
  <span
    className="fdsButton-label fdsButton--hidden"
  />
  <div
    className="alignChild--center--center fdsButton-icon--left fdsButton--hidden"
  >
    <Icon
      customSize={18}
    />
  </div>
  <div
    className="margin--left--half alignChild--center--center"
  >
    <CaretDownIcon
      customSize={12}
    />
  </div>
</Link>
`;
